packages/web-components/fast-foundation/src/date-picker/date-picker.ts

```diff
@@ -0,0 +1,906 @@
+     * Flag to prevent value setting more than once
+     */
+    private settingValue: boolean = false;
+
+    /**
+     * Validates and sets values
+     * @param values - object containing key-values to update
+     * @public
+     */
+    public setValue(values: {} | string | Date): void {
+        if (this.settingValue || !this.type) {
+            return;
+        }
+        this.settingValue = true;
+        this.textField?.control?.blur();
+        if (typeof values === "string" || values instanceof Date) {
+            values = this.getDateAsObject(values);
+        }
+        this.selectedDate = Object.assign({}, this.selectedDate, values);
+
+        if (this.type === "time") {
+            if (
+                !this.selectedDate.hour ||
+                this.selectedDate.minute === undefined ||
+                !this.selectedDate.meridian
+            ) {
+                this.settingValue = false;
+                return;
+            }
+
+            this.closeFlyout(true);
+        }
+
+        const selectedDate = Object.assign({}, this.getDateAsObject(), this.selectedDate);
+        const now = new Date();
+        const {
+            year = now.getFullYear(),
+            month = now.getMonth(),
+            day,
+            hour = now.getHours(),
+            minute,
+            meridian,
+        } = selectedDate;
+        const formatting = this.getFormatting();
+
+        let date = new Date(
+            year,
+            month - 1,
+            day,
+            hour + (meridian === "PM" ? 12 : 0),
+            minute
+        );
+
+        const start = new Date(this.min);
+        const end = new Date(this.max);
+        if (date.getTime() < start.getTime()) {
+            date = start;
+        } else if (date.getTime() > end.getTime()) {
+            date = end;
+        }
+
+        this.selectedDate = this.getDateAsObject(date);
+
+        const dateString: string = date.getTime()
+            ? this.dateFormatter.getDate(date, formatting)
+            : "";
+
+        this.value = dateString;
+        if (this.proxy instanceof HTMLInputElement) {
+            this.proxy.value = dateString;
+        }
+        if (this.textField) {
+            this.textField.control.value = dateString;
+        }
+
+        this.setViews();
+
+        if (this.type !== "datetime-local") {
+            this.closeFlyout(true);
+        }
+        this.settingValue = false;
+    }
+
+    /**
+     * Component has finished rendering
+     * @public
+     */
+    public connectedCallback(): void {
+        super.connectedCallback();
+        this.proxy.setAttribute("type", "text");
+        this.validate();
+        this.setCalendarTitle();
+        this.showCalendar = this.type.indexOf("date") >= 0;
+        this.showMonthPicker = this.type === "date" || this.type === "month";
+        this.showYearPicker = this.type === "month" || this.type === "year";
+
+        window.addEventListener("click", () => this.closeFlyout());
```

Comment:
Should we instead only add this event listener when the flyout is opened and remove it when it's closed?

Reply:
Good point. This would clean up the cases where the flyout isn't needed, readonly and disabled.